### PR TITLE
geotagging on map

### PR DESCRIPTION
**Pull request by phweyland**
allow to
- preview on map the different gpx tracks
- select meaningful tracks
- tune images datetime / offset to match some accurate positions on tracks (changing the date time / offset one sees the thumbs - in red - moving along the tracks).
- when the offset is ok one can lock it and extend the image selection
- apply datetime offset (if any) and position from gpx file

Other details:
- start date time is given in local time based on time zone. Should match with image date time. If not tune the offset.
- hovering start date time show the equivalent utc.
- the images column shows the number of images among the selected ones which match with the track. 

image

One of the first steps of this work, I've started adding some preview information on file chooser.
But after having the function on map I'm wondering if this is still useful.
image

Feedback and comments are welcome.

### CODE REVIEW

**Comment by TurboGit** [comment]
Looks interesting. I would have loved to have this when I was using a GPX track from my photo to geotag my pictures. Now I'm using Nikon SnapBridge which is connected with my Nikon camera and is used to record the GPS location directly into the RAW.

**Comment by phweyland** [comment]
> Now I'm using Nikon SnapBridge which is connected with my Nikon camera

That's simpler indeed ! My camera doesn't support this.

**Comment by johnny-bit** [comment]
Either way this is super cool :)

**Comment by phweyland** [comment]
The current ui is:
image

Apply GPX is now enabled only when date time offset is null (avoids mixed operations).
The bottom left button show / hide the images candidate to geotagging.
The bottom right button allows to only select the images candidate to geottaging. Afterwards the use can "apply offset" only on those. 


@dterrahe maybe you could help me here.

Using dt_ui_scroll_wrap with a treeview I have no issue until I show the columns header (see line 1724 in geotagging.c). I get these console messages:

```
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(darktable:47784): Gtk-CRITICAL **: 10:18:05.233: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(darktable:47784): Gtk-WARNING **: 10:18:05.233: gtk_widget_size_allocate(): attempt to allocate widget with width 12 and height -7

(darktable:47784): Gtk-WARNING **: 10:18:05.233: Negative content height -1 (allocation 1, extents 1x1) while allocating gadget (node scrollbar, owner GtkScrollbar)

(darktable:47784): Gtk-CRITICAL **: 10:18:05.233: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

```

**Comment by dterrahe** [comment]
Normally the scroll_wrapped widgets start with height 0 and get scaled up later, but in this case that means the first time there is no space for the header so it complains. Immediately after wrapping, asssign a minimum size request (that is at least enough for the header) and you are fine. This could go into dt_ui_scroll_wrap itself. Just an extra remark; you are saving both map.gpx_view and map.gpx_w in the gui data. This is probably not needed. Just remove gpx_w  everywhere and (un)hide gpx_view directly:
```
  GtkWidget *w = dt_ui_scroll_wrap(GTK_WIDGET(d->map.gpx_view), 100, "plugins/lighttable/geotagging/heighttracklist");
  gtk_widget_set_size_request(w, -1, 100);
  gtk_box_pack_start(GTK_BOX(self->widget), w, TRUE, TRUE, 0);
```
Couple of futher remarks after just glancing at the code and not testing much (so this is not a "review"):
- I don't have a test file + matching pictures and it seems if I load a random gpx I just get one line, so I can't completely test resizing. But you'd want to make the header and the list lines have equal height, otherwise you'll see partial lines. If that gets ugly, dt_ui_scroll_wrap may need some reworking to support headers explicitly.
- You are using GList a few times, but I'm not sure if you need its bidirectional features? (I'm just seeing g_list_next, not g_list_previous). Consider using GSList.
- Could the last last two lines below moved into the function definition? They repeat a few times.
```
  GtkWidget *box = _gui_init_datetime(&d->dt, 0, self);
  gtk_widget_set_halign(box, GTK_ALIGN_END);
  gtk_widget_set_hexpand(box, TRUE);
```

**Comment by phweyland** [comment]
Thanks for looking at this and for your explanations.
Outside the console messages that works fine but they are really ugly.
At the beginning I've used a grid but I haven't found the way to keep the headers in place when scrolling.
Thanks also for your comments on the code, I'll fix those.

**Comment by phweyland** [comment]
Yep, the error messages have gone. Thanks again.
Other changes done as well except GList (maybe useful is sorting needed).
Here is a gpx file with multiple track segments: [20110820.gpx.txt](https://github.com/darktable-org/darktable/files/5948050/20110820.gpx.txt)

**Comment by phweyland** [comment]
@TurboGit I'm done what I wanted to do with this PR. Should be ok to be reviewed. TIA.

File: src/common/gpx.c [hunk 1]
```diff
@@ -62,12 +60,20 @@ static GMarkupParser _gpx_parser
...
+  const dt_gpx_track_point_t *pb = (const dt_gpx_track_point_t *)b;
   glong diff = pa->time.tv_sec - pb->time.tv_sec;
   return diff != 0 ? diff : pa->time.tv_usec - pb->time.tv_usec;
 }
 
+static gint _sort_segment(gconstpointer a, gconstpointer b)
+{
+  const dt_gpx_track_segment_t *pa = (const dt_gpx_track_segment_t *)a;
+  const dt_gpx_track_segment_t *pb = (const dt_gpx_track_segment_t *)b;
+  glong diff = pa->start_dt.tv_sec - pb->start_dt.tv_sec;
```
Comment (TurboGit) [inline src/common/gpx.c:73]: const

File: src/libs/geotagging.c [hunk 2]
```diff
@@ -863,23 +1679,132 @@ void gui_init(dt_lib_module_t *self)
...
+  gtk_tree_view_append_column(GTK_TREE_VIEW(d->map.gpx_view), column);
+  column = _new_tree_text_column(_("images"), FALSE, 1.0, DT_GEO_TRACKS_IMAGES, PANGO_ELLIPSIZE_NONE);
+  gtk_tree_view_append_column(GTK_TREE_VIEW(d->map.gpx_view), column);
+
+  g_object_set(G_OBJECT(d->map.gpx_view), "has-tooltip", TRUE, NULL);
+  g_signal_connect(G_OBJECT(d->map.gpx_view), "query-tooltip", G_CALLBACK(_row_tooltip_setup), self);
+
+  // avoid ugly console pixman messages due to headers
+  gtk_tree_view_set_headers_visible(GTK_TREE_VIEW(d->map.gpx_view), FALSE);
+  GtkWidget *w = dt_ui_scroll_wrap(GTK_WIDGET(d->map.gpx_view), 100, "plugins/lighttable/geotagging/heighttracklist");
```
Comment (TurboGit) [inline src/libs/geotagging.c:1749]: The initial track list widget height is very small, less than one line. You need to add a preference into darktableconfig.xml.in with a minimal height value.

File: src/libs/geotagging.c [hunk 3]
```diff
@@ -863,23 +1679,132 @@ void gui_init(dt_lib_module_t *self)
...
+  d->map.preview_button = gtk_check_button_new_with_label(_("preview images"));
+  gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(d->map.preview_button), TRUE);
+  gtk_widget_set_sensitive(d->map.preview_button, FALSE);
+  gtk_label_set_ellipsize(GTK_LABEL(gtk_bin_get_child(GTK_BIN(d->map.preview_button))), PANGO_ELLIPSIZE_END);
+  gtk_grid_attach(grid, d->map.preview_button, 0, line, 1, 1);
+  gtk_widget_set_tooltip_text(GTK_WIDGET(d->map.preview_button),
+                              _("show on map images going to be geotagged"));
+  g_signal_connect(GTK_TOGGLE_BUTTON(d->map.preview_button), "toggled", G_CALLBACK(_images_preview_toggled), self);
+
+  d->map.select_button = dt_ui_button_new(_("select images"), _("select images going to be geotagged"), NULL);
```
Comment (TurboGit) [inline src/libs/geotagging.c:1767]: maybe just:

"select images to be geotagged"

File: src/libs/geotagging.c [hunk 4]
```diff
@@ -863,23 +1679,132 @@ void gui_init(dt_lib_module_t *self)
...
+  gtk_widget_set_sensitive(d->map.select_button, FALSE);
+  g_signal_connect(G_OBJECT(d->map.select_button), "clicked", G_CALLBACK(_select_images), self);
+  gtk_grid_attach(grid, d->map.select_button, 1, line, 1, 1);
+
+  _set_selected_images(self);
+  const int nb_sel = g_list_length(d->imgs);
+  char *nb = dt_util_dstrcat(NULL, "0/%d", nb_sel);
+  d->map.nb_imgs_label = dt_ui_label_new(nb);
+  gtk_widget_set_tooltip_text(GTK_WIDGET(d->map.nb_imgs_label),
+                              _("number of images going to be geotagged versus selected images"));
```
Comment (TurboGit) [inline src/libs/geotagging.c:1778]: again I would remove _going_.

File: src/libs/geotagging.c [hunk 5]
```diff
@@ -863,23 +1679,132 @@ void gui_init(dt_lib_module_t *self)
...
+  gtk_grid_set_column_spacing(grid, DT_PIXEL_APPLY_DPI(5));
+  line = 0;
+
+  d->map.preview_button = gtk_check_button_new_with_label(_("preview images"));
+  gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(d->map.preview_button), TRUE);
+  gtk_widget_set_sensitive(d->map.preview_button, FALSE);
+  gtk_label_set_ellipsize(GTK_LABEL(gtk_bin_get_child(GTK_BIN(d->map.preview_button))), PANGO_ELLIPSIZE_END);
+  gtk_grid_attach(grid, d->map.preview_button, 0, line, 1, 1);
+  gtk_widget_set_tooltip_text(GTK_WIDGET(d->map.preview_button),
+                              _("show on map images going to be geotagged"));
```
Comment (TurboGit) [inline src/libs/geotagging.c:1764]: "show images to be geotagged on map" ?

File: src/libs/geotagging.c [hunk 6]
```diff
@@ -863,23 +1679,132 @@ void gui_init(dt_lib_module_t *self)
...
+                             G_TYPE_INT, G_TYPE_INT, G_TYPE_INT, G_TYPE_STRING);
+
+  d->map.gpx_view = gtk_tree_view_new_with_model(GTK_TREE_MODEL(model));
+  g_object_unref(model);
+  gtk_widget_set_name(d->map.gpx_view, "gpx_list");
+  gtk_widget_set_tooltip_text(GTK_WIDGET(d->map.gpx_view),
+                              _("list of track segments in the gpx file, for each segment:"
+                                "\n- the start date time in local time (LT)"
+                                "\n- the number of track points"
+                                "\n- the number of images going to be geotagged "
```
Comment (TurboGit) [inline src/libs/geotagging.c:1722]: I'll remove _going_.

File: src/libs/geotagging.c [hunk 7]
```diff
@@ -133,13 +195,698 @@ static gboolean _lib_geotagging_filter_gpx(const GtkFileFilterInfo *filter_info,
...
+    dt_image_cache_read_release(darktable.image_cache, cimg);
+
+    dt_sel_img_t *img = g_malloc0(sizeof(dt_sel_img_t));
+    if(!img) continue;
+    memcpy(img->dt, dt, DT_DATETIME_LENGTH);
+    img->imgid = imgid;
+    d->imgs = g_list_prepend(d->imgs, img);
+  }
+  sqlite3_finalize(stmt);
+//  const GList *imgs = dt_view_get_images_to_act_on(TRUE, TRUE);
```
Comment (TurboGit) [inline src/libs/geotagging.c:877]: probably a left over of debug ?

**Review by TurboGit — CHANGES_REQUESTED** [review]
A first path on this PR.

Note that you've also said that the position during zoom in/out was not ok, I cannot confirm that on my side. Take the map and zoom out to see the whole world. If you zoom in by setting the mouse cursor in the center of the screen it is mostly ok, but set the mouse over say Dieppe and zoom in, you'll see that the mouse cursor is way below the initial location.

**Comment by TurboGit** [comment]
Also I have many errors on the console which I don't have without this PR:

```
(darktable:165695): GdkPixbuf-CRITICAL **: 09:44:39.126: gdk_pixbuf_new_from_data: assertion 'width > 0' failed

(darktable:165695): GLib-GObject-CRITICAL **: 09:44:39.126: g_object_set_qdata_full: assertion 'G_IS_OBJECT (object)' failed

(darktable:165695): GdkPixbuf-CRITICAL **: 09:44:39.126: gdk_pixbuf_scale_simple: assertion 'GDK_IS_PIXBUF (src)' failed

(darktable:165695): GLib-GObject-CRITICAL **: 09:44:39.126: g_object_unref: assertion 'G_IS_OBJECT (object)' failed

(darktable:165695): OsmGpsMap-WARNING **: 09:44:39.126: Error getting missing tile

(darktable:165695): GdkPixbuf-CRITICAL **: 09:44:39.126: gdk_pixbuf_new_from_data: assertion 'width > 0' failed

(darktable:165695): GLib-GObject-CRITICAL **: 09:44:39.126: g_object_set_qdata_full: assertion 'G_IS_OBJECT (object)' failed

(darktable:165695): GdkPixbuf-CRITICAL **: 09:44:39.126: gdk_pixbuf_scale_simple: assertion 'GDK_IS_PIXBUF (src)' failed

(darktable:165695): GLib-GObject-CRITICAL **: 09:44:39.126: g_object_unref: assertion 'G_IS_OBJECT (object)' failed

(darktable:165695): OsmGpsMap-WARNING **: 09:44:39.126: Error getting missing tile

(darktable:165695): GdkPixbuf-CRITICAL **: 09:44:39.131: gdk_pixbuf_new_from_data: assertion 'width > 0' failed

(darktable:165695): GLib-GObject-CRITICAL **: 09:44:39.131: g_object_set_qdata_full: assertion 'G_IS_OBJECT (object)' failed

(darktable:165695): GdkPixbuf-CRITICAL **: 09:44:39.131: gdk_pixbuf_scale_simple: assertion 'GDK_IS_PIXBUF (src)' failed

(darktable:165695): GLib-GObject-CRITICAL **: 09:44:39.131: g_object_unref: assertion 'G_IS_OBJECT (object)' failed

(darktable:165695): OsmGpsMap-WARNING **: 09:44:39.131: Error getting missing tile
```

**Comment by phweyland** [comment]
>  Take the map and zoom out to see the whole world. If you zoom in by setting the mouse cursor in the center of the screen it is mostly ok, but set the mouse over say Dieppe and zoom in, you'll see that the mouse cursor is way below the initial location.

I see. I knew I should not have missed to go to Dieppe... :smiley: 

> Also I have many errors on the console which I don't have without this PR:

I usually have these errors when the tiles are not yet in cache. Can you confirm you have still these errors having everything in cache ?

File: src/common/gpx.c (hunk 1 above)
Comment (phweyland) [reply src/common/gpx.c:73]: done

File: src/libs/geotagging.c (hunk 3 above)
Comment (phweyland) [reply src/libs/geotagging.c:1767]: slightly not the same meaning...
What I've set means (or at least it's my intention): images compatible with gpx file and which will be geotagged if one continues the process.
What you have written may not be accurate if there are more selected images than compatible images with the gpx file.

What do you think ? If you persist I'll do the change.

File: src/libs/geotagging.c (hunk 7 above)
Comment (phweyland) [reply src/libs/geotagging.c:877]: Correct. Removed.
Images_to_act_on() was the previous solution but creates glitches when moving the mouse around. 
Hovering is great in some circumstances but we should be able to switch it off when not desired.

File: src/libs/geotagging.c (hunk 3 above)
Comment (parafin) [reply src/libs/geotagging.c:1767]: "select images going to be geotagged" is not proper English I think, it should be "select images which are going to be geotagged" (if that what you meant), but to me it has exactly the same meaning as "select images to be geotagged". I don't understand the difference you're talking about. Do you mean "select images to try applying geotag to" (which sounds ugly)? I don't think this difference matters.

Comment (phweyland) [reply src/libs/geotagging.c:1767]: > "select images which are going to be geotagged" (if that what you meant)

Correct.
I do see a difference but my English is not accurate enough to discuss more about this.
I'll do the change.

**Comment by phweyland** [comment]
Thanks for your comments.
Requested changes are done, except the tiles errors.

Do we keep the file chooser track preview ?

**Comment by TurboGit** [comment]
> Do we keep the file chooser track preview ?

Yes, why not. I find this preview interesting !

**Comment by TurboGit** [comment]
> Requested changes are done, except the tiles errors.

Are you sure... I cannot reproduce now! Anyway, this looks ok to me. I'm merging. Thanks for all the work on the map view which was really lacking features.

**Pull request merged**